File: cpp/3484_Spreadsheet/Spreadsheet.cpp
#include <cassert>
#include <string>
#include <vector>

using namespace std;

class Spreadsheet {
 public:
  Spreadsheet(int rows) { this->grid = vector<vector<int>>(rows + 1, vector<int>(26)); }

  void setCell(string cell, int value) {
    auto [x, y] = getPos(cell);
    grid[x][y] = value;
  }

  void resetCell(string cell) {
    auto [x, y] = getPos(cell);
    grid[x][y] = 0;
  }

  int getValue(string formula) {
    int i = formula.find('+');
    string cell1 = formula.substr(1, i - 1);
    string cell2 = formula.substr(i + 1);
    return getCellVal(cell1) + getCellVal(cell2);
  }

 private:
  vector<vector<int>> grid;

  pair<int, int> getPos(const string &cell) {
    int x = stoi(cell.substr(1));
    int y = cell[0] - 'A';
    return {x, y};
  }

  int getCellVal(string &cell) {
    if (isalpha(cell[0])) {
      auto [x, y] = getPos(cell);
      return grid[x][y];
    } else {
      return stoi(cell);
    }
  }
};

int main() {
  Spreadsheet spreadsheet = Spreadsheet(3);
  assert(spreadsheet.getValue("=5+7") == 12);    // 返回 12 (5+7)
  spreadsheet.setCell("A1", 10);                 // 设置 A1 为 10
  assert(spreadsheet.getValue("=A1+6") == 16);   // 返回 16 (10+6)
  spreadsheet.setCell("B2", 15);                 // 设置 B2 为 15
  assert(spreadsheet.getValue("=A1+B2") == 25);  // 返回 25 (10+15)
  spreadsheet.resetCell("A1");                   // 重置 A1 为 0
  assert(spreadsheet.getValue("=A1+B2") == 15);  // 返回 15 (0+15)
}

File: rust/3484_Spreadsheet/Spreadsheet.rs
pub struct Spreadsheet {
    grid: Vec<Vec<i32>>,
}

impl Spreadsheet {
    pub fn new(rows: i32) -> Self {
        Spreadsheet {
            grid: vec![vec![0; 27]; (rows + 1) as usize],
        }
    }

    pub fn set_cell(&mut self, cell: String, value: i32) {
        let (x, y) = self.get_pos(&cell);
        self.grid[x][y] = value;
    }

    pub fn reset_cell(&mut self, cell: String) {
        let (x, y) = self.get_pos(&cell);
        self.grid[x][y] = 0;
    }

    pub fn get_value(&self, formula: String) -> i32 {
        let i = formula.find('+').unwrap();
        let cell1 = &formula[1..i];
        let cell2 = &formula[i + 1..];
        self.get_cell_val(cell1) + self.get_cell_val(cell2)
    }

    fn get_pos(&self, cell: &str) -> (usize, usize) {
        let x = cell[1..].parse::<usize>().unwrap();
        let y = cell.chars().next().unwrap() as usize - 'A' as usize;
        (x, y)
    }

    fn get_cell_val(&self, cell: &str) -> i32 {
        if cell.chars().next().unwrap().is_ascii_alphabetic() {
            let (x, y) = self.get_pos(cell);
            self.grid[x][y]
        } else {
            cell.parse().unwrap()
        }
    }
}

fn main() {
    let mut spread_sheet = Spreadsheet::new(3);
    assert_eq!(spread_sheet.get_value("=5+7".to_string()), 12);
    spread_sheet.set_cell("A1".to_string(), 10);
    assert_eq!(spread_sheet.get_value("=A1+6".to_string()), 16);
    spread_sheet.set_cell("B2".to_string(), 15);
    assert_eq!(spread_sheet.get_value("=A1+B2".to_string()), 25);
    spread_sheet.reset_cell("A1".to_string());
    assert_eq!(spread_sheet.get_value("=A1+B2".to_string()), 15);
}
